pydoctor/sphinx_ext/build_apidocs.py

```diff
@@ -33,26 +38,26 @@
 
 from pydoctor import __version__
 from pydoctor.driver import main, parse_args
-
+from pydoctor.sphinx_ext import get_source_reference
 
 logger = logging.getLogger(__name__)
 
+# Shared state between init and finish.
+_placeholders: Dict[str, str] = {}
```

Comment:
That was my first try... but I think it's a hack.

I think is better not to touch the Sphinx App or Config ...and mypy will also complain :) 

Even if the main deal is moved outside of sphinx, the placeholder injection thing will still be used in the sphinx extension.

Reply:
> That was my first try... but I think it's a hack.

Using global state is also a bit of a hack, since that assumes that there will be at most one build operating at the same time. I don't see a likely scenario in which that is not the case, but there is also no guarantee that there won't be multiple builds in the future.

> I think is better not to touch the Sphinx App or Config ...and mypy will also complain :)

Not touching `app` might be a good idea, since that is data that belongs to Sphinx. But the `config` object already contains attributes that are named by us such as `pydoctor_args`, so I don't think there is a problem adding attributes there.

To make mypy accept it, `Config.__setitem__()` can be used:
```py
    config['placeholders'] = placeholders
```

> Even if the main deal is moved outside of sphinx, the placeholder injection thing will still be used in the sphinx extension.

True, we still need it to inject `outdir`. However, `source_reference` is the only placeholder we currently have that is heavy to compute, so without that, it would also be an option to just reconstruct the placeholders in `on_build_finished()`. But it's possible we'll get placeholders in the future that are non-trivial to compute and don't belong in pydoctor itself, so storing the placeholders is probably the better solution.